lib/logidze/meta.rb

```diff
@@ -29,12 +30,65 @@ def initialize(meta, &block)
       end
 
       def perform
-        return if block.nil?
+        raise ArgumentError, "Block must be given" unless block
         return block.call if meta.nil?
 
-        ActiveRecord::Base.transaction { call_block_in_meta_context }
+        call_block_in_meta_context
       end
 
+      def current_meta
+        meta_stack.reduce(:merge) || {}
+      end
+
+      def meta_stack
+        Thread.current[:meta] ||= []
+        Thread.current[:meta]
+      end
+
+      def encode_meta(value)
+        connection.quote(ActiveSupport::JSON.encode(value))
+      end
+
+      def pg_reset_meta_param(prev_meta)
+        if prev_meta.empty?
+          pg_clear_meta_param
+        else
+          pg_set_meta_param(prev_meta)
+        end
+      end
+    end
+
+    class MetaWithTransaction < MetaWrapper # :nodoc:
+      private
+
+      def call_block_in_meta_context
+        ActiveRecord::Base.transaction do
```

Comment:
Let's do the following:
- Implement `call_block_in_meta_context` in the base class:

```ruby
def call_block_in_meta_context
  prev_meta = current_meta

  meta_stack.push(meta)

  pg_set_meta_param(current_meta)
  result = block.call
  result
ensure
  pg_reset_meta_param(prev_meta)
  meta_stack.pop
end
```

- Here we can override this method by wrapping `super` with transaction:

```ruby
def call_block_in_meta_context
  ActiveRecord::Base.transaction { super }
end
```

- Drop `call_block_in_meta_context` in `MetaWithoutTransaction` (the current implementation is not complete — we should take meta_stack into account; the base implementation above would work in non-transactional case without any change)